Rework ARC MDB runner

evgeniy-paltsev opened the pull request:
mdb runner is quite special as it can be used to run Zephyr on both simulator (nSIM) and real hardware.
As of today it is really misleading as same command (west flash) will run Zephyr in simulation for one board and try to run it on HW for another board. Things are getting worse for boards supporting both runs in simulation and on real hardware.

This PR 
 * improves overall MDB runner UX by splitting it for mdb-hw (for runs on HW) and mdb-nsim (for runs in simulation) runners. 
 * fixes nsim_hs_smp run testing in sanitycheck [uses mdb-nsim]
 * allows to run ARC nSIM boards (starting with nsim_ prefix) Zephyr in both simulation and on real HW [uses mdb-nsim and mdb-hw]

Commit:
ARC: west: split mdb runner for mdb-hw & mdb-nsim runners

mdb runner is quite special as it can be used to run Zephyr on
both simulator (nSIM) and real hardware.
However it is really misleading as same command (west flash)
will run Zephyr in simulation for one board and try to run it
on HW for another board. Things are getting worse for boards
supporting both runs in simulation and on real hardware.

Let's split mdb runner for mdb-hw (for runs on HW) and mdb-nsim
(for runs in simulation) runners.

Signed-off-by: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>

Commit:
ARC: west: mdb runner: record MDB's CLD process pid to file

Record MDB's CLD process pid to file so this process can be
terminated by sanitycheck infrastructure.
Update mdb runners test to be able to handle changes.

Signed-off-by: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>

Commit:
sanitycheck: ARC: nsim: allow to run tests on HW as well

ARC nSIM boards (starting with nsim_ prefix) allow to run
tests in simulator (nSIM). However with proper west runner
the tests can be run on real HW (FPGA based).

So check device_testing flag firstly and rollback to nSIM
simulation run in case of device_testing is not set.

Signed-off-by: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>

Commit:
sanitycheck: ARC: mdb-nsim: add missing stuff for simulation run

Add missing infrastructure for simulation run on mdb-nsim.
When the nSIM via mdb simulation support was added to Zephyr
scripts the handler call was lost. So nSIM via mdb simulation
was broken from the beginning.

Fix that bu adding missing handler call fro mdb-nsim.
Note that we use BinaryHandler which call west flash. It's
required in case of nSIM via mdb as we create mdb.pid file with
mdb-nsim runner help.

Signed-off-by: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>

abrodkin commented on scripts/west_commands/runners/mdb.py:
```diff
@@ -18,6 +20,29 @@
 def simulation_run(mdb_runner):
     return mdb_runner.nsim_args != ''
 
+def get_cld_pid(mdb_process):
```
IMHO would be good to add some background info in the commit message on why that PID mess is needed.
I.e. that MDB spawns that CLD process which handles real work and that just killing MDB doesn't affect CLD process, thus we need to reach it directly by its PID.

evgeniy-paltsev replied:
We have comment with some info in the place were we call get_cld_pid. However I've added more detailed description and move it to the place where we define get_cld_pid.

abrodkin reviewed (approved):
Good that we have much better working MDB-related stuff and indeed may now run the same binaries in both simulation and on matching FPGA platform.

ruuddw commented on .github/workflows/west_cmds.yml:
```diff
@@ -55,7 +55,7 @@ jobs:
       run: |
         pip3 install wheel
-        pip3 install pytest west pyelftools canopen progress mypy intelhex
+        pip3 install pytest west pyelftools canopen progress mypy intelhex psutil
```
psutil doesn't exist on Windows, don't think this can be made a generic dependency.

evgeniy-paltsev replied:
`psutil` exists on Windows, here is proof from official documentation:
https://psutil.readthedocs.io/en/latest/#about

Moreover there is west command test performed bu github workfow with windows:
https://github.com/evgeniy-paltsev/zephyr/runs/1249826436
It was run from my branch with this change.

ruuddw replied:
missed that, thanks for showing the proof!

ruuddw commented:
Great to have support for nsim_hs_smp in sanity check!
Not fully sure if I like the split of mdb into mdb-nsim and mdb-hw as seperate runners. From a user/conceptual point of view, I think there are boards (including simulated boards), and there are runners. For me 'mdb' is a single runner, requiring different arguments depending on what board is used -> can't there be a single mdb runner, and the board/target details are made part of either some intelligence in the mdb runner, or defined in a standard argument file located with the boards?

mbolivar-nordic reviewed (commented):
Hi there ARC maintainers and developers, I just want you to know that I see this PR but I'm waiting for you all to come to consensus on the arch specific portions before reviewing the Python bits in the runners package.

Commit:
ARC: nsim: enable mdb-hw runner for nsim_* boards

ARC nSIM boards (starting with nsim_ prefix) allow to run
Zephyr in simulator and on real hardware.
Allow to run Zephyr on HW by enabling mdb-hw runner.

Signed-off-by: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>

nashif commented on scripts/sanity_chk/sanitylib.py:
```diff
@@ -336,6 +336,7 @@ def __init__(self, instance, type_str):
         super().__init__(instance, type_str)
 
         self.terminated = False
+        self.call_west_flash = False
```
should this be in the Handler class instead, alongside `self.call_make_run`

evgeniy-paltsev replied:
Sure, will move to Handler class.

evgeniy-paltsev replied:
Done

nashif commented on scripts/sanity_chk/sanitylib.py:
```diff
@@ -336,6 +336,7 @@ def __init__(self, instance, type_str):
         super().__init__(instance, type_str)
 
         self.terminated = False
+        self.call_west_flash = False
```
AFAIK, `west flash` for the purpose of running a binary on the host works only with nsim, for example we can't call `west flash` on qemu or renode.

evgeniy-paltsev replied:
I was under impression that we can call west flash on qemu as we have `scripts/west_commands/runners/qemu.py`. But I've checked it and it appears that it is only a stub.

I am wondering why do we have it?

ruuddw replied:
For ARC, this has historical reasons. As we didn't support flashing of our development boards (EMSK e.g. doesn't have flash), the 'flash' target was implemented by downloading and running the executable instead of flash&run. For nSIM this then was copied.

Should a 'west run' be introduced to be more consistent, instead of the 'make run'?
Using mdb in 'make run' isn't straightforward, as I understood from Evgeniy.

mbolivar-nordic replied:
> I am wondering why do we have it?

Various people keep meaning to make it not a stub :). You can remove it if you really want but I don't think it does any harm. It would be nice to have it, just nobody has had time to do it.

ruuddw commented:
ok, if introducing 'west flash' runner is fine I have no further concerns. Just one request: some documentation for the mdb runners would be good to add, e.g. mention the mdb-nsim one in the nSIM board documentation, and the mdb-hw for relevant boards as well. Or, introduce the mdb runners and sanitycheck support as an addition to the generic MetaWare toolchain documentation.

carlescufi commented:
@mbolivar-nordic I think this should be ready for a Python-level review now

Commit:
ARC: boards: nsim: update documentation

Update nsim board documentation:
 * add info about run on HW (HAPS)
 * update info about dependencies in case of single / multi core
   runs in simulation and run on HW

Signed-off-by: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>

evgeniy-paltsev commented:
@ruuddw 
I've updated the nsim board documentation. I added info about runs on hardware (HAPS), multi core nsim runs, tools required, etc. 

I didn't add specific info about runners to nsim documentation as it all available with generated board-specific help which can be viewed with `west flash --context` command.

ruuddw reviewed (approved):
Good to be merged I think. Thanks for updating the docs with the explanation on HAPS fpga and nSIM usage as targets for development, with corresponding mdb-hw and mdb-nsim runners.

carlescufi commented:
> Good to be merged I think. Thanks for updating the docs with the explanation on HAPS fpga and nSIM usage as targets for development, with corresponding mdb-hw and mdb-nsim runners.

I'd like @mbolivar-nordic to take a look first.
